cli/file_watcher.rs

```diff
@@ -28,51 +90,55 @@ async fn error_handler(watch_future: WatchFuture) {
 }
 
 pub async fn watch_func<F>(
-  watch_paths: &[PathBuf],
+  paths: &[PathBuf],
   closure: F,
 ) -> Result<(), AnyError>
 where
   F: Fn() -> WatchFuture,
 {
+  let (_watcher, receiver) = new_watcher(paths)?;
+  let mut debounce = Debounce::new(receiver, DEBOUNCE_TIMEOUT_MS);
   loop {
     let func = error_handler(closure());
     let mut is_file_changed = false;
     select! {
-      _ = file_watcher(watch_paths) => {
-          is_file_changed = true;
-          info!(
-            "{} File change detected! Restarting!",
-            colors::intense_blue("Watcher")
-          );
-        },
-      _ = func => { },
-    };
+      _ = debounce.next() => {
+        is_file_changed = true;
+        info!(
+          "{} File change detected! Restarting!",
+          colors::intense_blue("Watcher"),
+        );
+      },
+      _ = func => {},
+    }
     if !is_file_changed {
       info!(
         "{} Process terminated! Restarting on file change...",
-        colors::intense_blue("Watcher")
+        colors::intense_blue("Watcher"),
       );
-      file_watcher(watch_paths).await?;
+      debounce.next().await;
       info!(
         "{} File change detected! Restarting!",
-        colors::intense_blue("Watcher")
+        colors::intense_blue("Watcher"),
       );
     }
   }
 }
 
-pub async fn file_watcher(paths: &[PathBuf]) -> Result<(), AnyError> {
-  let (sender, mut receiver) =
-    mpsc::channel::<Result<NotifyEvent, AnyError>>(16);
-  let sender = std::sync::Mutex::new(sender);
+fn new_watcher(
+  paths: &[PathBuf],
+) -> Result<
+  (RecommendedWatcher, Receiver<Result<NotifyEvent, AnyError>>),
+  AnyError,
+> {
+  let (sender, receiver) = mpsc::channel::<Result<NotifyEvent, AnyError>>();
 
   let mut watcher: RecommendedWatcher =
     Watcher::new_immediate(move |res: Result<NotifyEvent, NotifyError>| {
       let res2 = res.map_err(AnyError::from);
-      let mut sender = sender.lock().unwrap();
       // Ignore result, if send failed it means that watcher was already closed,
       // but not all messages have been flushed.
-      let _ = sender.try_send(res2);
+      let _ = sender.send(res2);
     })?;
```

Comment:
I just came up with an idea: instead of using `mpsc`, update `Debounce.last_event` inside the callback of `Watcher::new_immediate`, and `Debounce::poll_next` should just check if `DEBOUNCE_TIMEOUT_MS` elapsed since the last event.